src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,950 @@
+
+
+	if (entry == NULL || entry->dpe_val_ptr == NULL) {
+		D_GOTO(out, rc);
+	}
+
+	/* convert acl to list of strings */
+	acl = (struct daos_acl *)entry->dpe_val_ptr;
+	rc = daos_acl_to_strs(acl, &acl_strs, &len_acl);
+	if (rc != 0) {
+		D_ERROR("failed to convert acl to strs "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+	attr_dims[0] = len_acl;
+	attr_dtype = H5Tcopy(H5T_C_S1);
+	if (attr_dtype < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Tset_size(attr_dtype, H5T_VARIABLE);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, acl_strs);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attributes "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	for (i = 0; i < len_acl; i++) {
+		D_FREE(acl_strs[i]);
+	}
+	D_FREE(acl_strs);
+
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static int
+serialize_str(hid_t file_id, struct daos_prop_entry *entry,
+	      const char *prop_str)
+{
+
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype = 0;
+	hid_t	attr_dspace = 0;
+	hid_t	usr_attr = 0;
+
+	if (entry == NULL || entry->dpe_str == NULL) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_C_S1);
+	if (attr_dtype < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Tset_size(attr_dtype, strlen(entry->dpe_str) + 1);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Tset_strpad(attr_dtype, H5T_STR_NULLTERM);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype, attr_dspace,
+			      H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		rc = -DER_MISC;
+		D_GOTO(out, rc);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, entry->dpe_str);
+	if (status < 0) {
+		rc = -DER_IO;
+		D_ERROR("failed to write attribute "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	/* close hdf5 objects */
+	if (usr_attr > 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype > 0)
+		H5Tclose(attr_dtype);
+	if (attr_dspace > 0)
+		H5Sclose(attr_dspace);
+	return rc;
+}
+
+static int
+serialize_uint(hid_t file_id,
+	       struct daos_prop_entry *entry,
+	       const char *prop_str)
+{
+	int	rc = 0;
+	hid_t	status = 0;
+	hsize_t	attr_dims[1];
+	hid_t	attr_dtype = 0;
+	hid_t	attr_dspace = 0;
+	hid_t	usr_attr = 0;
+
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_NATIVE_UINT64);
```

Comment:
Do you need to copy that int? you can use H5T_NATIVE_UINT64 directly in H5Acreate(), no?

Reply:
@mchaarawi I did some probably not best things when I was writing this and learning HDF5...and yes it does work to just directly pass it to H5Acreate instead 👍